implement tutorial gtag

e111077 opened the pull request:
Fires tutorial start and end events. Also sets `document.title` which is also useful for gtag events

```
tutorial_start: {
  category: tutorials,
  label: tutorial folder name,
  value: step tutorial page was loaded,
}

tutorial_end {
  category: tutorials,
  label: tutorial folder name,
}
```

## tutorial_start event

fired when the page and manifest area loaded, and includes the step the page was loaded on.

## tutorial_end event

fired when the next button is pressed and the page index matches the length of steps in the laoded manifest -1. Signaling progression

## faults

- start is fired whenever the page loads so that may include reloads
  - can likely be deduplicated by session ID cloud-side
  - hopefully can fire automated event if session has start index of 0 and also fires end event
- if initial page loaded is the final page, then next button cannot be pressed unless they go back and press right
  - doesn't signal progression so this is fine, but may have a mismatched start to end

AndrewJakubowicz commented on packages/lit-dev-content/src/components/litdev-tutorial.ts:
```diff
@@ -753,6 +773,23 @@ export class LitDevTutorial extends LitElement {
     };
   }
+
+  private reportMetrics(idx: number, tutorialUrl: string) {
+    if (idx === 0 || !hasRecordedStart) {
+      window.gtag?.('event', 'tutorial_start', {
+        category: 'tutorials',
+        event_label: tutorialUrl,
+        value: idx,
+      });
+      hasRecordedStart = true;
+    } else if (idx === this._manifest.steps.length - 1 && !hasRecordedEnd) {
+      window.gtag?.('event', 'tutorial_end', {
+        category: 'tutorials',
+        event_label: tutorialUrl,
+      });
+      hasRecordedEnd = true;
+    }
+  }
```
Is it worth splitting out the gtag related logic into a separate analytics file, and making this a pure function that isn't bound to the component?

AndrewJakubowicz reviewed (commented):
This is awesome! I tested with https://tagassistant.google.com/ and everything looks great.

Two nitpicks:
 - Can we split and document the analytics code separate from the component. Even if just in the same file?
 - Is it worth adding an extra event like `tutorial_progress` that triggers at each step with the value of the step? Then we can see if certain tutorials have a drop off point?

aomarks commented on packages/lit-dev-content/src/components/litdev-tutorial.ts:
```diff
@@ -753,6 +773,23 @@ export class LitDevTutorial extends LitElement {
+        value: idx,
+      });
+      hasRecordedStart = true;
+    } else if (idx === this._manifest.steps.length - 1 && !hasRecordedEnd) {
```
Shouldn't we record each step between start and end, too?

aomarks commented on packages/lit-dev-content/src/types/gtag.d.ts:
```diff
@@ -0,0 +1,7 @@
+interface Window {
```
License header

e111077 commented:
WOW TIL @AndrewJakubowicz! Thanks so much for that link!

Moved the gtag logic to its own file but still requires a local private method to track state.

Added licenses.

Added `tutorial_progress` event

Possible confusions for progress is that if they go back, it won't record back progress but it will record if they go back and then right. WDYT? Should It also record back?

aomarks commented on packages/lit-dev-content/src/components/litdev-tutorial.ts:
```diff
@@ -774,20 +789,17 @@ export class LitDevTutorial extends LitElement {
+        hasRecordedEnd = true;
+        break;
+      case 'tutorial_progress':
+      default:
```
A `never` check would be good. Otherwise the `default` case doesn't need to be here.

aomarks commented on packages/lit-dev-content/src/util/gtag-helpers.ts:
```diff
@@ -0,0 +1,50 @@
+  | 'tutorial_start'
+  | 'tutorial_progress'
+  | 'tutorial_end'
+  | null;
```
```suggestion
  | null;

```

aomarks commented on packages/lit-dev-content/src/components/litdev-tutorial.ts:
```diff
@@ -58,6 +62,9 @@ interface ExpandedTutorialStep {
 
 type CheckStatus = 'INDETERMINATE' | 'CHECKING' | 'PASSED' | 'FAILED';
 
+let hasRecordedStart = false;
```
This is a module scoped variable, so does it assume new tutorials are always loaded with a full page reload? Seems like that could be a little brittle, since a refactoring to make this a little more SPAish could break that assumption.

AndrewJakubowicz reviewed (approved):
Small code nits. Pre-approving. The analytic events are working great!

e111077 commented on packages/lit-dev-content/src/components/litdev-tutorial.ts:
```diff
@@ -300,6 +307,13 @@ export class LitDevTutorial extends LitElement {
     return this;
   }
 
+  update(changed: PropertyValues<this>): void {
+    if (this._manifest.header) {
+      document.title = `${this._manifest.header} Tutorial – Lit`;
+    }
+    super.update(changed);
+  }
```
It doesn't really matter because it's not like this is run on the server. Personally I think earlier in the lifecycle is better

e111077 commented on packages/lit-dev-content/src/util/gtag-helpers.ts:
```diff
@@ -0,0 +1,50 @@
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+export type TutorialMetricsOptions = {
```
lol nice catch! this one was evolved from a bunch of `'string'|'string2'`

e111077 commented:
Question: should I be recording progress when they hit the back button? Currently only broadcasting forward

AndrewJakubowicz commented on packages/lit-dev-content/src/util/gtag-helpers.ts:
```diff
@@ -24,21 +24,26 @@ export const reportTutorialMetrics = ({
   numSteps,
 }: TutorialMetricsOptions): TutorialMetricEvent => {
+  // A user can load a tutorial from a URL at a step that is not 0, hence
+  // we shoul consider whether start has not been recorded already.
```
Nit: "should"

AndrewJakubowicz commented:
I continue to happily LGTM this PR! No strong opinion on back button. Wouldn't it technically be inferable from multiple duplicate progress steps provided the user doesn't hit the end?

aomarks commented:
Sorry for delay
